Index: xfce4-mpc-plugin.c
===================================================================
--- xfce4-mpc-plugin.c	(révision 1458)
+++ xfce4-mpc-plugin.c	(copie de travail)
@@ -233,28 +233,19 @@
 
    xfce_panel_plugin_block_menu (plugin);
 
-   dlg = gtk_dialog_new_with_buttons (_("Properties"),
-                                       GTK_WINDOW (gtk_widget_get_toplevel
-                                                   (GTK_WIDGET (plugin))),
-                                       GTK_DIALOG_DESTROY_WITH_PARENT |
-                                       GTK_DIALOG_NO_SEPARATOR,
-                                       GTK_STOCK_CLOSE, GTK_RESPONSE_OK,
-                                       NULL);
+   dlg = xfce_titled_dialog_new_with_buttons (_("Mpd Client Plugin"),
+                                              NULL,
+                                              GTK_DIALOG_NO_SEPARATOR,
+                                              GTK_STOCK_CLOSE,
+                                              GTK_RESPONSE_OK,
+                                              NULL);
+   xfce_titled_dialog_set_subtitle (XFCE_TITLED_DIALOG (dlg), _("Properties"));
 
+   gtk_window_set_position   (GTK_WINDOW (dlg), GTK_WIN_POS_CENTER);
+   gtk_window_set_icon_name  (GTK_WINDOW (dlg), "xfce-multimedia");
+
    g_signal_connect (dlg, "response", G_CALLBACK (mpc_dialog_response), dialog);
 
-   gtk_container_set_border_width (GTK_CONTAINER (dlg), 2);
-
-   pb = xfce_themed_icon_load ("xfce4-panel", 48);
-   gtk_window_set_icon (GTK_WINDOW (dlg), pb);
-   g_object_unref (pb);
-
-   header = xfce_create_header (NULL, _("Mpd Client Plugin"));
-   gtk_widget_set_size_request (GTK_BIN (header)->child, -1, 32);
-   gtk_container_set_border_width (GTK_CONTAINER (header), 5);
-   gtk_widget_show (header);
-   gtk_box_pack_start (GTK_BOX (GTK_DIALOG (dlg)->vbox), header, FALSE, TRUE, 0);
-
    vbox = gtk_vbox_new (FALSE, 8);
    gtk_container_set_border_width (GTK_CONTAINER (vbox), 5);
    gtk_widget_show (vbox);
