diff --git a/configure.ac b/configure.ac
index 28bf56f..76d3e24 100644
--- a/configure.ac
+++ b/configure.ac
@@ -6,7 +6,7 @@ PACKAGE=Ario
 VERSION=0.2
 
 AC_INIT(ario,ario_version)
-AC_CONFIG_HEADER(config.h)
+AM_CONFIG_HEADER(config.h)
 AM_INIT_AUTOMAKE($PACKAGE,$VERSION)
 
 AC_PROG_CC
diff --git a/m4/.empty b/m4/.empty
index e69de29..7b4d68d 100644
--- a/m4/.empty
+++ b/m4/.empty
@@ -0,0 +1 @@
+empty
\ No newline at end of file
